File: psana_examples/tags/V00-01-04/src/DumpCsPad.cpp
//--------------------------------------------------------------------------
// File and Version Information:
// 	$Id$
//
// Description:
//	Class DumpCsPad...
//
// Author List:
//      Andrei Salnikov
//
//------------------------------------------------------------------------

//-----------------------
// This Class's Header --
//-----------------------
#include "psana_examples/DumpCsPad.h"

//-----------------
// C/C++ Headers --
//-----------------
#include <algorithm>
#include <iomanip>
#include <iterator>

//-------------------------------
// Collaborating Class Headers --
//-------------------------------
#include "MsgLogger/MsgLogger.h"
#include "psddl_psana/cspad.ddl.h"

//-----------------------------------------------------------------------
// Local Macros, Typedefs, Structures, Unions and Forward Declarations --
//-----------------------------------------------------------------------

using namespace psana_examples;
PSANA_MODULE_FACTORY(DumpCsPad)

//		----------------------------------------
// 		-- Public Function Member Definitions --
//		----------------------------------------

namespace psana_examples {

//----------------
// Constructors --
//----------------
DumpCsPad::DumpCsPad (const std::string& name)
  : Module(name)
{
  m_src = configStr("source", "DetInfo(:Cspad)");
  m_src2x2 = configStr("source2x2", "DetInfo(:Cspad2x2)");
}

//--------------
// Destructor --
//--------------
DumpCsPad::~DumpCsPad ()
{
}

// Method which is called at the beginning of the calibration cycle
void 
DumpCsPad::beginCalibCycle(Event& evt, Env& env)
{
  MsgLog(name(), trace, "in beginCalibCycle()");

  shared_ptr<Psana::CsPad::ConfigV1> config1 = env.configStore().get(m_src);
  if (config1.get()) {
    
    WithMsgLog(name(), info, str) {
      str << "CsPad::ConfigV1:";
      str << "\n  concentratorVersion = " << config1->concentratorVersion();
      str << "\n  runDelay = " << config1->runDelay();
      str << "\n  eventCode = " << config1->eventCode();
      str << "\n  inactiveRunMode = " << config1->inactiveRunMode();
      str << "\n  activeRunMode = " << config1->activeRunMode();
      str << "\n  tdi = " << config1->tdi();
      str << "\n  payloadSize = " << config1->payloadSize();
      str << "\n  badAsicMask0 = " << config1->badAsicMask0();
      str << "\n  badAsicMask1 = " << config1->badAsicMask1();
      str << "\n  asicMask = " << config1->asicMask();
      str << "\n  quadMask = " << config1->quadMask();
      str << "\n  numAsicsRead = " << config1->numAsicsRead();
      str << "\n  numQuads = " << config1->numQuads();
      str << "\n  numSect = " << config1->numSect();
    }
    
  }

  shared_ptr<Psana::CsPad::ConfigV2> config2 = env.configStore().get(m_src);
  if (config2.get()) {
    
    WithMsgLog(name(), info, str) {
      str << "CsPad::ConfigV2:";
      str << "\n  concentratorVersion = " << config2->concentratorVersion();
      str << "\n  runDelay = " << config2->runDelay();
      str << "\n  eventCode = " << config2->eventCode();
      str << "\n  inactiveRunMode = " << config2->inactiveRunMode();
      str << "\n  activeRunMode = " << config2->activeRunMode();
      str << "\n  tdi = " << config2->tdi();
      str << "\n  payloadSize = " << config2->payloadSize();
      str << "\n  badAsicMask0 = " << config2->badAsicMask0();
      str << "\n  badAsicMask1 = " << config2->badAsicMask1();
      str << "\n  asicMask = " << config2->asicMask();
      str << "\n  quadMask = " << config2->quadMask();
      str << "\n  numAsicsRead = " << config2->numAsicsRead();
      str << "\n  numQuads = " << config2->numQuads();
      str << "\n  numSect = " << config2->numSect();
      str << "\n  roiMask =";
      for (unsigned i = 0; i < config2->numQuads(); ++ i) {
        str.setf(std::ios::showbase);
        str << ' ' << std::hex << config2->roiMask(i) << std::dec;
      }
      str << "\n  numAsicsStored =";
      for (unsigned i = 0; i < config2->numQuads(); ++ i) {
        str << ' ' << config2->numAsicsStored(i);
      }
    }
    
  }

  shared_ptr<Psana::CsPad::ConfigV3> config3 = env.configStore().get(m_src);
  if (config3.get()) {
    
    WithMsgLog(name(), info, str) {
      str << "CsPad::ConfigV3:";
      str << "\n  concentratorVersion = " << config3->concentratorVersion();
      str << "\n  runDelay = " << config3->runDelay();
      str << "\n  eventCode = " << config3->eventCode();
      str << "\n  protectionEnable = " << config3->protectionEnable();
      str << "\n  protectionThresholds:";
      for (unsigned i = 0; i < config3->numQuads(); ++ i) {
        str << "\n    adcThreshold=" << config3->protectionThresholds(i).adcThreshold()
            << " pixelCountThreshold=" << config3->protectionThresholds(i).pixelCountThreshold();
      }
      str << "\n  inactiveRunMode = " << config3->inactiveRunMode();
      str << "\n  activeRunMode = " << config3->activeRunMode();
      str << "\n  tdi = " << config3->tdi();
      str << "\n  payloadSize = " << config3->payloadSize();
      str << "\n  badAsicMask0 = " << config3->badAsicMask0();
      str << "\n  badAsicMask1 = " << config3->badAsicMask1();
      str << "\n  asicMask = " << config3->asicMask();
      str << "\n  quadMask = " << config3->quadMask();
      str << "\n  numAsicsRead = " << config3->numAsicsRead();
      str << "\n  numQuads = " << config3->numQuads();
      str << "\n  numSect = " << config3->numSect();
      str << "\n  roiMask =";
      for (unsigned i = 0; i < config3->numQuads(); ++ i) {
        str.setf(std::ios::showbase);
        str << ' ' << std::hex << config3->roiMask(i) << std::dec;
      }
      str << "\n  numAsicsStored =";
      for (unsigned i = 0; i < config3->numQuads(); ++ i) {
        str << ' ' << config3->numAsicsStored(i);
      }
    }
    
  }
}

// Method which is called with event data
void 
DumpCsPad::event(Event& evt, Env& env)
{

  shared_ptr<Psana::CsPad::DataV1> data1 = evt.get(m_src);
  if (data1.get()) {
    
    WithMsgLog(name(), info, str) {
      str << "CsPad::DataV1:";
      int nQuads = data1->quads_shape()[0];
      for (int q = 0; q < nQuads; ++ q) {
        const Psana::CsPad::ElementV1& el = data1->quads(q);
        str << "\n  Element #" << q ;
        str << "\n    virtual_channel = " << el.virtual_channel() ;
        str << "\n    lane = " << el.lane() ;
        str << "\n    tid = " << el.tid() ;
        str << "\n    acq_count = " << el.acq_count() ;
        str << "\n    op_code = " << el.op_code() ;
        str << "\n    quad = " << el.quad() ;
        str << "\n    seq_count = " << el.seq_count() ;
        str << "\n    ticks = " << el.ticks() ;
        str << "\n    fiducials = " << el.fiducials() ;
        str << "\n    frame_type = " << el.frame_type() ;
        str << "\n    sb_temp = [ ";
        const uint16_t* sb_temp = el.sb_temp();
        std::copy(sb_temp, sb_temp+Psana::CsPad::ElementV1::Nsbtemp, 
            std::ostream_iterator<uint16_t>(str, " "));
        str << "]";
        str << "\n    data_shape = [ ";
        const std::vector<int> dshape = el.data_shape();
        std::copy(dshape.begin(), dshape.end(), std::ostream_iterator<int>(str, " "));
        str << "]";
      }

    }
  }


  shared_ptr<Psana::CsPad::DataV2> data2 = evt.get(m_src);
  if (data2.get()) {
    
    WithMsgLog(name(), info, str) {
      str << "CsPad::DataV2:";
      int nQuads = data2->quads_shape()[0];
      for (int q = 0; q < nQuads; ++ q) {
        const Psana::CsPad::ElementV2& el = data2->quads(q);
        str << "\n  Element #" << q ;
        str << "\n    virtual_channel = " << el.virtual_channel() ;
        str << "\n    lane = " << el.lane() ;
        str << "\n    tid = " << el.tid() ;
        str << "\n    acq_count = " << el.acq_count() ;
        str << "\n    op_code = " << el.op_code() ;
        str << "\n    quad = " << el.quad() ;
        str << "\n    seq_count = " << el.seq_count() ;
        str << "\n    ticks = " << el.ticks() ;
        str << "\n    fiducials = " << el.fiducials() ;
        str << "\n    frame_type = " << el.frame_type() ;
        str << "\n    sb_temp = [ ";
        const uint16_t* sb_temp = el.sb_temp();
        std::copy(sb_temp, sb_temp+Psana::CsPad::ElementV2::Nsbtemp, 
            std::ostream_iterator<uint16_t>(str, " "));
        str << "]";
        str << "\n    data_shape = [ ";
        const std::vector<int> dshape = el.data_shape();
        std::copy(dshape.begin(), dshape.end(), std::ostream_iterator<int>(str, " "));
        str << "]";
      }

    }
  }

  shared_ptr<Psana::CsPad::MiniElementV1> mini1 = evt.get(m_src2x2);
  if (mini1.get()) {

    WithMsgLog(name(), info, str) {
      str << "CsPad::MiniElementV1:";
      str << "\n  virtual_channel = " << mini1->virtual_channel() ;
      str << "\n  lane = " << mini1->lane() ;
      str << "\n  tid = " << mini1->tid() ;
      str << "\n  acq_count = " << mini1->acq_count() ;
      str << "\n  op_code = " << mini1->op_code() ;
      str << "\n  quad = " << mini1->quad() ;
      str << "\n  seq_count = " << mini1->seq_count() ;
      str << "\n  ticks = " << mini1->ticks() ;
      str << "\n  fiducials = " << mini1->fiducials() ;
      str << "\n  frame_type = " << mini1->frame_type() ;
      str << "\n  sb_temp = [ ";
      const uint16_t* sb_temp = mini1->sb_temp();
      std::copy(sb_temp, sb_temp+Psana::CsPad::ElementV2::Nsbtemp,
          std::ostream_iterator<uint16_t>(str, " "));
      str << "]";
      str << "\n  data_shape = [ ";
      const std::vector<int> dshape = mini1->data_shape();
      std::copy(dshape.begin(), dshape.end(), std::ostream_iterator<int>(str, " "));
      str << "]";
    }
  }

}

} // namespace psana_examples

File: psana_examples/tags/jbarrera-12-05-15/src/DumpCsPad.py
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id: dump_simple.py 2622 2011-11-11 14:35:00Z salnikov@SLAC.STANFORD.EDU $
#
# Description:
#  Class DumpCsPad
#
# Author List:
#  Joseph S. Barrera III
#------------------------------------------------------------------------

"""Example module for accessing SharedIpimb data.

This software was developed for the LCLS project.  If you use all or 
part of it, please give an appropriate acknowledgment.

@version $Id: dump_simple.py 2622 2011-11-11 14:35:00Z salnikov@SLAC.STANFORD.EDU $

@author Joseph S. Barrera III
"""

#------------------------------
#  Module's version from SVN --
#------------------------------
__version__ = "$Revision: 2622 $"
# $Source$

#--------------------------------
#  Imports of standard modules --
#--------------------------------
import sys
import logging

#-----------------------------
# Imports for other modules --
#-----------------------------
from pypdsdata import xtc

#----------------------------------
# Local non-exported definitions --
#----------------------------------

# local definitions usually start with _

#---------------------
#  Class definition --
#---------------------
class DumpCsPad(object) :
    """Class whose instance will be used as a user analysis module. """

    #----------------
    #  Constructor --
    #----------------
    def __init__ (self, source=""):
        self.m_src = source

    #-------------------
    #  Public methods --
    #-------------------
    def beginCalibCycle(self, evt, env):

        print "in beginCalibCycle()"

        print "jobName=", env.jobName()
        print "instrument=", env.instrument()
        print "experiment=", env.experiment()
        print "expNum=", env.expNum()
        print "calibDir=", env.calibDir()

        print "self.m_src = %s" % (self.m_src)

        self.source = env.configStr("source", "DetInfo(:Cspad)")

        print "configV1 =", env.getConfig("Psana::CsPad::ConfigV1", self.source)
        print "configV2 =", env.getConfig("Psana::CsPad::ConfigV2", self.source)
        print "configV3 =", env.getConfig("Psana::CsPad::ConfigV3", self.source)

        config = env.getConfig("Psana::CsPad::ConfigV3", self.source)
        if not config:
            config = env.getConfig("Psana::CsPad::ConfigV2", self.source)
        if not config:
            config = env.getConfig("Psana::CsPad::ConfigV1", self.source)
        if not config:
            print "unrecoginized CsPad::Config version"
            sys.exit(1)

        print "config type is ", type(config)


        version = config.concentratorVersion()
        print "version=", version
        s = "CsPad::ConfigV1:"
        s = s + "\n  concentratorVersion = " + str(config.concentratorVersion())
        s = s + "\n  runDelay = " + str(config.runDelay())
        s = s + "\n  eventCode = " + str(config.eventCode())
        s = s + "\n  inactiveRunMode = " + str(config.inactiveRunMode())
        s = s + "\n  activeRunMode = " + str(config.activeRunMode())
        s = s + "\n  tdi = " + str(config.tdi())
        s = s + "\n  payloadSize = " + str(config.payloadSize())
        s = s + "\n  badAsicMask0 = " + str(config.badAsicMask0())
        s = s + "\n  badAsicMask1 = " + str(config.badAsicMask1())
        s = s + "\n  asicMask = " + str(config.asicMask())
        s = s + "\n  quadMask = " + str(config.quadMask())
        s = s + "\n  numAsicsRead = " + str(config.numAsicsRead())
        s = s + "\n  numQuads = " + str(config.numQuads())
        s = s + "\n  numSect = " + str(config.numSect())
        """
        if hasattr(config, "protectionEnable"):
            s = s + "\n  protectionEnable = " + str(config.protectionEnable())
        if hasattr(config, "roiMask"):
            s = s + "\n  roiMask = " + str(config.roiMask())
        if hasattr(config, "numAsicsStored"):
            s = s + "\n  numAsicsStored = " + str(config.numAsicsStored())
        """
        print s

    def event(self, evt, env):
        evt_data = evt.getData("Psana::CsPad::DataV2", self.source)
        if not evt_data:
            evt_data = evt.getData("Psana::CsPad::DataV1", self.source)
            if not evt_data:
                print "data is of unexpected format"
                sys.exit(1)
        print evt_data

        s = "CsPad::DataV1:"
        print "evt_data = ", evt_data
        print "evt_data.quads_shape() = ", evt_data.quads_shape()
        nQuads = evt_data.quads_shape()[0]
        for q in range(nQuads):
            el = evt_data.quads(q)

            s = s + "\n  Element #" + str(q)
            s = s + "\n    virtual_channel = " + str(el.virtual_channel())
            s = s + "\n    lane = " + str(el.lane())
            s = s + "\n    tid = " + str(el.tid())
            s = s + "\n    acq_count = " + str(el.acq_count())
            s = s + "\n    op_code = " + str(el.op_code())
            s = s + "\n    quad = " + str(el.quad())
            s = s + "\n    seq_count = " + str(el.seq_count())
            s = s + "\n    ticks = " + str(el.ticks())
            s = s + "\n    fiducials = " + str(el.fiducials())
            s = s + "\n    frame_type = " + str(el.frame_type())

            # str << "\n    sb_temp = [ ";
            # const ndarray<uint16_t, 1>& sb_temp = el.sb_temp();
            # std::copy(sb_temp.begin(), sb_temp.end(), std::ostream_iterator<uint16_t>(str, " "));
            # str << "]";
            sb_temp = el.sb_temp()
            print "sb_temp = ", sb_temp
            print "type(sb_temp) = ", type(sb_temp)
            print "sb_temp.size = ", sb_temp.shape[0]
            s = s + "\n    sb_temp = [ "
            for i in range(sb_temp.shape[0]):
                s = s + str(sb_temp[i]) + " "
            s = s + "]"
            print s # XXX

            # const ndarray<int16_t, 3>& data = el.data();
            # str << "\n    data_shape = [ ";
            # std::copy(data.shape(), data.shape()+3, std::ostream_iterator<int>(str, " "));
            # str << "]";
            data = el.data()
            print "data.ndim = ", data.ndim
            s = s + "\n    data_shape = [ "
            for i in range(data.ndim):
                s = s + str(data.shape[i]) + " "
            s = s + "]"
            print s # XXX

            # str << "\n    common_mode = [ ";
            # for (unsigned i = 0; i != data.shape()[0]; ++ i) {
            #   str << el.common_mode(i) << ' ';
            # }
            # str << "]";
            s = s + "\n    common_mode = [ "
            for i in range(data.shape[0]):
                print "i=", i
                print "el.common_mode(i)=", el.common_mode(i)
                s = s + str(el.common_mode(i)) + " "
            s = s + "]"
            print s # XXX

            # str << "\n    data = [";
            # for (unsigned s = 0; s != data.shape()[0]; ++ s) {
            #   str << "\n        [ ";
            #   for (unsigned i = 0; i < 10; ++ i) str << data[s][0][i] << ' ';
            #   str << "... ]";
            # }
            # str << "\n    ]";
            s = s + "\n    data = ["
            for i in range(data.shape[0]):
                s = s + "\n        [ "
                for j in range(10):
                    s = s + str(data[i][0][j]) + " "
                s = s + "... ]"
            s = s + "\n    ]"
            print s

    def endJob(self, evt, env):
        pass
